### ISSUE TITLE
Defining @Header for all controller method at one place

### ISSUE THREAD

**Issue by gardelin** [lukeautry/tsoa#1357]
I know there is @Header decorator for specifing request header

`public async getUser(@Path() id: string, @Header('X-Access-Token') _token: string) {}`

To avoid putting `@Header('X-Access-Token') _token: string` in every controller method is there possibility to define it only once.

Something like this:

```
@Header('X-Access-Token)
export class UserController extends BaseController {
    ...
    public async getUser(@Path() id: string) {}
    ...
}
```

**Comment by WoH** [comment]
Hi, that sounds like a good feature, would you like to submit a PR that adds this?

**Comment by gardelin** [comment]
@WoH I would like to submit a PR, I tried to code but after few hours I am still lost haha. Any chance you point me to right way?

Let's say that we have @Header decorator defined like this.

```
@Header('X-Access-Token, {parameterName: 'token', type: 'string'}) 
export class UserController extends BaseController {
    ...
    public async getUser(@Path() id: string) {}
    ...
}
```

My guess is that in `ControllerGenerator` class I need to get and parse @Header decorator and in `buildMethods()` forward it to `MethodGenerator` class. Then in `MethodGenerator` class in `Generate` method merge it with results of `buildParameters()` .

My biggest problem is how to convert `@Header('X-Access-Token, {parameterName: 'token', type: 'string'}) ` to `Tsoa.Parameter`;

Also is this right way to define header parameterName and type. Should user be able to define multiple default headers?

```
@Header('X-Access-Token-1, {parameterName: 'token1', type: 'string'}) 
@Header('X-Access-Token,-2 {parameterName: 'token2', type: 'string'}) 

export class UserController extends BaseController {
    ...
    public async getUser(@Path() id: string) {}
    ...
}
```

What do you think?

**Comment by gardelin** [comment]
@WoH I would like to submit a PR, I tried to code but after few hours I am still lost haha. Any chance you point me to right way?

Let's say that we have @Header decorator defined like this.

```
@Header('X-Access-Token, {parameterName: 'token', type: 'string'}) 
export class UserController extends BaseController {
    ...
    public async getUser(@Path() id: string) {}
    ...
}
```

My guess is that in `ControllerGenerator` class I need to get and parse @Header decorator and in `buildMethods()` forward it to `MethodGenerator` class. Then in `MethodGenerator` class in `Generate` method merge it with results of `buildParameters()` .

My biggest problem is how to convert `@Header('X-Access-Token, {parameterName: 'token', type: 'string'}) ` to `Tsoa.Parameter`;

Also is this right way to define header parameterName and type. Should user be able to define multiple default headers?

```
@Header('X-Access-Token-1, {parameterName: 'token1', type: 'string'}) 
@Header('X-Access-Token,-2 {parameterName: 'token2', type: 'string'}) 

export class UserController extends BaseController {
    ...
    public async getUser(@Path() id: string) {}
    ...
}
```

What do you think?

**Comment by WoH** [comment]
First, I think I'd look to make the API work slightly differently:

```ts
class Ctrl {
  @Header()
  public one: string

  @Get()
  public async ....
}
```

Then, in the controllerGenerator, track that and change the templates to, before calling the req handler, set the property.

**Comment by jackey8616** [comment]
@gardelin 
Is this feature request is still proceeding?
I'm willing to take this.  
Or do you need any help from me?

**Comment by kvbutler** [comment]
Yeah I would like similar functionality, for example define a generic BaseController that has implementations getAll<Array<T>> and getById<T> method can we do something like that currently? Would like methods and models to be created without overrides in the child class.
